### PR TITLE
Fix false positive test passes

### PR DESCRIPTION
**Pull request by natemcmaster**
The end to end tests for dotnet-watch were passing when they should not have been due to race conditions in the test.

Bugs: 
https://github.com/aspnet/DotNetTools/issues/212
https://github.com/aspnet/DotNetTools/issues/224

cc @moozzyk

### CODE REVIEW

**Comment by natemcmaster** [comment]
Ping. @moozzyk or @cesarbs could you review?

File: test/Microsoft.DotNet.Watcher.Tools.FunctionalTests/AwaitableProcess.cs [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        private void StartProcessingOutput(StreamReader streamReader)
+        {
+            _source = _source ?? new BufferBlock<string>();
+            Interlocked.Increment(ref _reading);
+            Task.Run(() =>
+            {
+                var buffer = new char[1];
+                var sb = new StringBuilder();
+                var lastChar = '\0';
+                while (streamReader.Read(buffer, 0, 1) > 0)
```
Comment (moozzyk) [inline test/Microsoft.DotNet.Watcher.Tools.FunctionalTests/AwaitableProcess.cs]: why not just ReadLine?

File: test/Microsoft.DotNet.Watcher.Tools.FunctionalTests/AwaitableProcess.cs [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+                    }
+                }
+            }
+
+            return null;
+        }
+
+        private void StartProcessingOutput(StreamReader streamReader)
+        {
+            _source = _source ?? new BufferBlock<string>();
```
Comment (moozzyk) [inline test/Microsoft.DotNet.Watcher.Tools.FunctionalTests/AwaitableProcess.cs]: for some reason I feel uneasy with `BufferBlock` - feels like something from a different world...

File: test/Microsoft.DotNet.Watcher.Tools.FunctionalTests/TestProjects/GlobbingApp/Program.cs [hunk 3]
```diff
@@ -3,37 +3,19 @@
...
-        private static void Block()
-        {
-            while (true)
-            {
-                ConsoleWrite("Blocked...");
-                Thread.Sleep(1000);
-            }
+            Console.WriteLine("Started");
+            Console.WriteLine("PID = " + Process.GetCurrentProcess().Id);
+            Console.WriteLine("Defined types = " + typeof(Program).GetTypeInfo().Assembly.DefinedTypes.Count());
```
Comment (moozzyk) [inline test/Microsoft.DotNet.Watcher.Tools.FunctionalTests/TestProjects/GlobbingApp/Program.cs]: See it now in a test..

**Comment by cesarbs** [comment]
Everything looks ok to me, but this is a fairly large change. Can you elaborate on what exactly was wrong with the tests?

File: test/Microsoft.DotNet.Watcher.Tools.FunctionalTests/NoDepsAppTests.cs [hunk 4]
```diff
@@ -2,132 +2,63 @@
...
-                    File.WriteAllText(fileToChange, programCs);
-
-                    wait.Wait(_defaultTimeout,
-                        expectedToChange: true,
-                        errorMessage: $"Process did not restart because {scenario.StartedFile} was not changed");
-                }
-            }
+            await _app.StartWatcher().OrTimeout();
+            var pid = await _app.GetProcessId().OrTimeout();
+            await _app.HasExited().OrTimeout(); // process should exit after run
```
Comment (moozzyk) [inline test/Microsoft.DotNet.Watcher.Tools.FunctionalTests/NoDepsAppTests.cs]: Remind me - the process is not running after this line until the file is changed when watcher will restart the process and let it exit - is this correct?

**Comment by natemcmaster** [comment]
The tests were designed to use files on disk for IPC, but this kind of IPC had race conditions.

Example:
1. Xunit starts dotnet-watch which will run test-app
2. test-app writes to a file 'started.txt'
3. xunit waits for changes to 'started.txt' to change. When it does, it attempts to change a file and see if dotnet-watch will restart test-app.

The problem with this approach was step 3 used a polling file watcher. Depending on when the poll happens, 'started.txt' could represent the first time test-app runs, or the second time test-app has run.
One solution could have been to add additional waits and checks to ensure the change to 'started.txt' represented what we wanted, but this got messy.

These changes simplify the whole test design by instead just reading stdout from the dotnet-watch process in order to determine what is happening inside this process.

**Comment by cesarbs** [comment]
Cool. 🚢 when Travis is green.

File: test/Microsoft.DotNet.Watcher.Tools.FunctionalTests/NoDepsAppTests.cs (hunk 4 above)
Comment (natemcmaster) [inline test/Microsoft.DotNet.Watcher.Tools.FunctionalTests/NoDepsAppTests.cs]: Yes, that's right. If the process dotnet-watch is monitoring exits on its own, dotnet-watch will still hang around and wait for file changes before it restarts the process.

File: test/Microsoft.DotNet.Watcher.Tools.FunctionalTests/AwaitableProcess.cs (hunk 2 above)
Comment (natemcmaster) [inline test/Microsoft.DotNet.Watcher.Tools.FunctionalTests/AwaitableProcess.cs]: It seemed like a good way to do async producer/consumer. Are there potential issues with this API?

File: test/Microsoft.DotNet.Watcher.Tools.FunctionalTests/TestProjects/GlobbingApp/Program.cs (hunk 3 above)
Comment (natemcmaster) [inline test/Microsoft.DotNet.Watcher.Tools.FunctionalTests/TestProjects/GlobbingApp/Program.cs]: yeah, it's meant to help tests inspect if the app recompiled with(out) the file changes.

**Pull request merged**